docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,841 @@
+    worker: {
+      dev: {
+        createEnvironment: createWorkerEnvironment,
+      },
+    },
+  },
+})
+```
+
+:::
+
+`RemoteRunnerTransport` and `RemoteEnvironmentTransport` are meant to be used together. If you don't use either of them, then you can define your own function to communicate between the runner and the server.
+
+## ModuleRunnerHMRConnection
+
+**Type Signature:**
+
+```ts
+export interface ModuleRunnerHMRConnection {
+  /**
+   * Checked before sending messages to the client.
+   */
+  isReady(): boolean
+  /**
+   * Send a message to the client.
+   */
+  send(message: string): void
+  /**
+   * Configure how HMR is handled when this connection triggers an update.
+   * This method expects that the connection will start listening for HMR updates and call this callback when it's received.
+   */
+  onUpdate(callback: (payload: HMRPayload) => void): void
+}
+```
+
+This interface defines how HMR communication is established. Vite exports `ServerHMRConnector` from the main entry point to support HMR during Vite SSR. The `isReady` and `send` methods are usually called when the custom event is triggered (like, `import.meta.hot.send("my-event")`).
+
+`onUpdate` is called only once when the new module runner is initiated. It passed down a method that should be called when connection triggers the HMR event. The implementation depends on the type of connection (as an example, it can be `WebSocket`/`EventEmitter`/`MessageChannel`), but it usually looks something like this:
+
+```js
+function onUpdate(callback) {
+  this.connection.on('hmr', (event) => callback(event.data))
+}
+```
+
+The callback is queued and it will wait for the current update to be resolved before processing the next update. Unlike the browser implementation, HMR updates in a module runner will wait until all listeners (like, `vite:beforeUpdate`/`vite:beforeFullReload`) are finished before updating the modules.
+
+## Environments during build
+
+Plugin hooks also receive the environment instance during build. This replaces the `ssr` boolean we have been passing them so far.
+
+In the CLI, calling `vite build` will build the Client. It is equivalent to calling `vite build --environment=client`.
+
+The build the SSR server, the `--ssr` shourcut can be used: `vite build --ssr`. This is equivalent to calling `vite build --environment=ssr`.
+
+Other non-default environments can be build using `vite build --environment=name`.
+
+## Building all environments
+
+Calling `vite build --all` will instantiate a `ViteBuilder` (build-time equivalent to a `ViteDevServer`) to build all configured environments for production. By default the build of environments is run in series respecting the order of the `environments` array. A framework or user can further configure how the build tasks are excecuted using:
+
+```js
+export default {
+  builder: {
+    runBuildTasks: asnyc (builder, buildTasks) => {
+      return Promise.all(buildTasks.map( task => task.run() ))
```

Comment:
It may be a bit difficult to do so given that the core parts heavily depend on the  environment abstraction that I don't know if Rollup could adopt. It is an interesting idea though, maybe once this API is stable and has proven useful, Rolldown could have environments as a first class citizen too. But this starts to blur the lines a bit of responsibilities. We always thought of Rollup/esbuild/rolldown working at a single environment level, and Vite is the abstraction that coordinates (client/ssr). So initially I'm leaning to keep that distinction.

Note: If rolldown wants to implements the dev plugin container, then it is kind of inevitable. But I'm unsure of this strategy. Maybe if we move the pluginContainer to Rust, it should still live in the Vite repo.

Reply:
Reconsidering it, I agree with that 👍